### ISSUE TITLE
Star/label photo's in focused view

### ISSUE THREAD

**Issue by hampelman1** [photoprism/photoprism#713]
I would love to be able to label and (most importantly) star (5-star system) my photo's in the focused view mode..

This way I can use photoprism to manage my large collection...

**Comment by lastzero** [comment]
Are you actually using 1 to 5 stars in a way that makes a significant difference if an image has 4 or 5 stars?

I felt like this provides no value to me (actually huge waste of time) and most photographers, so we decided to implement a simple like button instead. It would be relatively easy to map 3-5, 4-5 or just 5 stars to this so that existing metadata can be used.

**Comment by hampelman1** [comment]
yes, the numbers do make a difference. I have a huge collection, and by using the star system, i can filter them somewhat. I could maybe do with 3...but then compatability wth other tools (and metadata fields???) is broken..

Synology Photostation uses the 5 star system, as does digikam. (https://docs.kde.org/trunk5/en/extragear-graphics/digikam/using-dam.html > scroll to rating/ranking)

1 star: Meh. Photo could safely be deleted in the future (for example when running out of space)
2 stars: Photo should be retained, but nothing special
3 stars: good quality photo, cloud be printed in photobook (if you're into that*)
4 stars: highlights of a trip/year: use in a slideshow for others
5 stars: very good photo. Top of the top. Poster-like quality.

* my wife likes very large photobooks. Distinction between ratings may vary upon personal workflow...

**Comment by graciousgrey** [comment]
I changed the title to keep this ticket to implement a star rating. 
We have another ticket to customise actions that can be done in fullscreen mode --> https://github.com/photoprism/photoprism/issues/508

**Comment by bernsel** [comment]
I also use the Star Rating system, as a way to quickly categorize images. In DigiKam on MacOS, it is possible to use keyboard shortcuts (Command+1 to Command+5) to very quickly set a Star Rating on a photo when you preview or scroll through, which makes it a quick and effective way of organizing images. I do not use the stars as a quality metric, but rather as a classification metric. This is how I apply them.

1 star:  Picture to be deleted
2 stars: Work-related picture - should be moved to work folder
3 stars: Picture of me + daughter, or me+girlfriend
4 stars: Image good enough to include in Album
5 stars: Image I want to print as Poster

After classifying like this in DigiCam, I can filter on the exact number of stars in DigiCam, as a quick double-check to see I got each classification right. After that, I select "write metadata to files" in DigiKam, to preserve the rating in the files themselves (and delete the 1-star images in bulk).

Arguably, I could use 5 different user-defined Tags instead, to accomplish the same thing. But Stars are convenient because they have shortcuts in many photo apps, not just in DigiKam, but also for example Appollo One and Lyn. Only some apps allow allocating custom keyboard shortcuts to set user-defined Tags. - so the star rating approach feels like a more universal approach for quick classification.

**Comment by lastzero** [comment]
Thanks for your feedback & ideas!

**Comment by PescheHelfer** [comment]
I would also like to support this idea. Currently, I use Adobe Bridge to organize my foto libary, which does support 5 star rating using keyboard shortcuts. I would like to switch to Photoprism, but 5-star rating is a killer argument for me.

**Comment by sebw** [comment]
I'm also interested in the feature in the sense that the Canon 5D mark IV (and a few other models) cameras have a physical "rate" button that I use quite extensively. I know for sure that Rawtherapee imports the rating.

![image](https://user-images.githubusercontent.com/2285094/130353956-e41b9d59-50c2-4a4d-99a3-ab0eb9a76420.png)

Having the ratings imported in Photoprism would be quite awesome.

**Comment by charles-997** [comment]
A question on the implementation here:
Would it be any easier to implement the ability to easily view the 1-5 star rating in Photoprism but not necessarily edit said metadata?
As others have stated I rate my pics in adobe using 1-5 star system and I'd assume many people use photoprism as only a very nice UI to view the pics but not to make edits or interact with them.

Just a thought, no clue the ease of implementation and definitely aware the team has bigger fish to fry at this time!

**Comment by graciousgrey** [comment]
> Would it be any easier to implement the ability to easily view the 1-5 star rating in PhotoPrism but not necessarily edit said metadata?

Not really, 5 star ratings in general are not a complicated feature (compared to other tasks). Probably the most time consuming part is a nice UI that works on desktop and mobile :)

**Comment by adambeck7** [comment]
As a supporter I'd love to see this implemented as I've used the 5 star system in both Darktable and Lightroom in a similar manner as @bernsel.  

From an end-user perspective, I'd expect/like to see it with the metadata on desktop and down at the near the location data on mobile. Something like this on desktop:

<img width="308" alt="Screen Shot 2022-09-17 at 11 01 40 AM" src="https://user-images.githubusercontent.com/18033909/190868606-c19f11d3-75ce-4d52-b64c-807679e104a7.png">


And this on mobile:
<img width="308" alt="Screen Shot 2022-09-17 at 11 08 40 AM" src="https://user-images.githubusercontent.com/18033909/190868755-44e7e745-25db-4d52-be8d-168efcfe2aa5.PNG">

Thanks for all your hard work on this!

**Comment by stevealexr** [comment]
> And this on mobile:
> <img alt="Screen Shot 2022-09-17 at 11 08 40 AM" width="308" src="https://user-images.githubusercontent.com/18033909/190868755-44e7e745-25db-4d52-be8d-168efcfe2aa5.PNG">

The photo viewer is implemented using photoswipe library. The caption is added as html string in `addCaptionHTMLFn` function. If I add the vuetify v-rating component directly in the template, the image will overlap with the rating and I can't figure out other ways to add the rating component.

**Comment by fvsion** [comment]
Would love to see this implemented and a filter option added, it would greatly simplify sorting my library and make albums wayyyy easier to create.

Would certainly prefer it to edit the rating metadata, as I often use photoprism to go through my images and delete the junk, then sync the backend to lightroom/darktable for editing. Perhaps a toggle in the setting for ratings to be sidecar vs. metadata updates. This might become necessary once multiple users are implemented and one users rating should not affect that of another assuming they have access to the same backend photos. Just some thoughts.

**Comment by raidenashish** [comment]
I would like to implement it. Will Like to contribute to this issue.

**Comment by lastzero** [comment]
We would love if you could help us with any of the issues labeled "Help Wanted", as this is a core feature that we would like to implement ourselves, if we have time and get help with the other issues.

**Comment by Danilogomes76** [comment]
z

**Comment by anarcat** [comment]
Darktable also has a 5-star system for rating photos, and also a "rejected" state. i use "rejected" for photos that can be deleted, everything else is a spectrum... typically i rate one-star for documentation photo that have no artistic interest, for example...

**Comment by fracklaus** [comment]
Any progress here? This feature is really needed to keep photoprism compatible with other photo software like digikam, darktable, etc.

**Comment by graciousgrey** [comment]
Unfortunately, we haven't had time to work on this yet. Our [roadmap](https://github.com/orgs/photoprism/projects/5) shows what tasks are currently in progress and what will be implemented next.

To show us that a feature is important to you, simply leave a :+1:  in the issue description. 
Thank you very much :heart:

**Comment by heitorPB** [comment]
There's an Exif metadata `Rating`, will it be used here to import the star rating from a photo? Darktable adds this metadata to the photo (if enabled, of course):

```bash
$ exiv2 -p a pr _DSC7081.jpg | grep Rating
Exif.Image.Rating                            SLong       1  3
Exif.Photo.ISOSpeedRatings                   Short       1  800
Xmp.xmp.Rating                               XmpText     1  3
```

**Comment by graciousgrey** [comment]
@heitorPB Once we have support for 5 star ratings we plan to also read this information from various metadata fields.

**Comment by fred1701d** [comment]
Hi. I set the ratings on my photos according to my own "feeling". 3 stars is what is worth showing. 4 stars is what is worth printing and showing first. 5 star is exceptional. 1 and 2 starts are either duplicates or should be thrown away. 
So I'd like to have a way of searching for 4 stars or more". If there was a filter is all views, it would be even better but that's more difficult to implement. 
Just a filter on the ratin in search view would be a good start.

**Comment by c0f** [comment]
@fred1701d I do the same, 5 star ratings are a subjective but extremely useful organisational tool. I use them in Digikam and scripts that use ratings to copy photos for family, my slideshow server and to my Photoprism container.

**Comment by Tilmann-Glaeser** [comment]
I am waiting for this feature as well. But please store this information in the sidecar and not in the image file directly. See: [issue 1092](https://github.com/photoprism/photoprism/discussions/1092#discussioncomment-8547219).

**Comment by fvsion** [comment]
> I am waiting for this feature as well. But please store this information in the sidecar and not in the image file directly. See: [issue 1092](https://github.com/photoprism/photoprism/discussions/1092#discussioncomment-8547219).

This would defeat the use case of applying these ratings in an app like lightroom/darktable and having it reflected in Photoprism...or conversely using photoprism to casually cull images by rating and have that rating show up in lightroom. 

Perhaps a sidecar toggle would be good, but everyone else asking for this feature is requesting it go in the exif data specifically because they "rate in camera" or use it to cull images as mentioned above. 

It's definitely worth discussing the use cases and intended purpose by supports of this feature.

**Comment by Tilmann-Glaeser** [comment]
@fvsion good point for storing the information in the exif in the image. Maybe it could be implemented as
- store metadata (if possible) always in the sidecar
- store metadata (if possible) always in the image
- merge the sidecar metadata into the image when _exported_
a challenge is how to deal with the metadata in the image if defined in the sidecar as well. Which source has the priority?

I do see that the best seems to be storing all metadata in the image itself. So I guess I have to chose a different backup definitions (in terms of retention time / backup generations) for photos. Or has anyone a good idea?

**Comment by alelom** [comment]
Just restating what people say here -- definitely a needed feature! I've been considering Photoprism for my NAS but I really need this feature (in particular, for video files) to do the switch.

**Comment by larryg215** [comment]
This would be a fantastic addition!

**Comment by Bluemeus** [comment]
I also would like this feature, to sort and organize my huge collection. I use this for my pictures with the eos 80d in Lightroom and this is quite the best and fastest way to get through so much photos. I also include all of our smartphone pictures via AutoSync and so everything in one could be accessed, which is absolutely perfect for us!

**Comment by brownphotographic** [comment]
Hi Photoprism team,

Could we get this one onto the roadmap? I imported my photo collection (with star ratings) and would love to be able to see those, filter against them, and also edit them from the PhotoPrism UI

**Comment by cimba007** [comment]
Seems to have aged quie a bit. Would love to see this feature!

**Comment by anarcat** [comment]
one thing i have immediately needed today is a way to "flag" some images as more important than others. i'm using ratings for this right now, like i'd do a shoot and anything above and including (say) two stars get shared with the family or client or whatever. if there would be a way to add an exif/sidecar rating that could be recognized in photoprism after, and imported, and could be used to make an album that's shared after, that would already be fantastic.

in other words: how does the *existing* "star" "label" work exactly? is it imported from an existing EXIF tag? that could just be enough for me...

or, in general, are labels recognized when importing images?

update: ah, nevermind, image labels are not imported either, see https://github.com/photoprism/photoprism/issues/1143

**Comment by kagahd** [comment]
I'm sorry to say but not having xmp 1-5 star rating makes PhotoPrism **pointless** for me. As said by others in this ongoing thread already (since 2020!) , the xmp Rating field is used by so many widely used photo apps (Microsoft, XnView, Digikam, LightRoom, PiGallery2, F-Stop, etc.) that I couldn't believe, until I found this thread, that xmp Ratings are really missing in such a feature packed app like PhotoPrism. That's really a pity and should be changed as soon as possible!

**Comment by fvsion** [comment]
@kagahd and at this development team...I've basically stopped using your program....and I would be a paying customer, but photoprism is useless without this, otherwise it's a pretty web app to show off photos. A pretty web app that I have to spend time managing to make sure DNG/JPG stack properly, and constantly have to regenerate the library because of the way trash is implemented. So close...yet soooooooo very far.

**Comment by kagahd** [comment]
@fvsion I agree what you said. I also don't use PhotoPrism because of this missing feature.
I was and I am still very happy with [PiGallery2](https://github.com/bpatrik/pigallery2) but since it's a read-only gallery, I just implemented [pigallery2-metadata-editor](https://github.com/kagahd/pigallery2-metadata-editor) which allows to edit the tags I'm interested in. It can be easily extended to support more tags and it leaves the PiGallery2 source completely untouched, so you benefit from future updates without merge conflicts. 
Thanks to this extension, I'm sticking with PiGallery2.

**Comment by aktentasche** [comment]
Nice to read that this is moved onto the roadmap 

I actually just started usign the five star workflow to pre-sort (and delete) my images before they go into photoprism and my software creates .xmp containing the star rating along with the RAW files which are then imported.

So I would like to suggest to use those .xmp files to store/retrieve the star rating as it seems to be also some kind of industry standard.

I am not sure how photoprism works but I would assume some of the metadata is copied to an internal database on import, so it would be nice if the rating that came with an image in form of an .xmp file would be imported as well. Only problem might be that at some point you might get a desynchronisation between the star rating in the .xmp file and the database (if someone choses to change a rating) but I would say that's a death worth dying.

**Comment by fvsion** [comment]
> Nice to read that this is moved onto the roadmap
> 
> I actually just started usign the five star workflow to pre-sort (and delete) my images before they go into photoprism and my software creates .xmp containing the star rating along with the RAW files which are then imported.
> 
> So I would like to suggest to use those .xmp files to store/retrieve the star rating as it seems to be also some kind of industry standard.
> 
> I am not sure how photoprism works but I would assume some of the metadata is copied to an internal database on import, so it would be nice if the rating that came with an image in form of an .xmp file would be imported as well. Only problem might be that at some point you might get a desynchronisation between the star rating in the .xmp file and the database (if someone choses to change a rating) but I would say that's a death worth dying.

This is two fold. DNG files themselves can store star ratings in their metadata, so when you rate a photo in camera that data is saved within the file. Many software suites (Lightroom, darktable, etc.) can either modify the DNG file OR they usually have settings to store the data in what is largely a standardized XMP file, Photoprism should implement both and allow the user to decide which one takes precedence. Both workflows are really required for this to be implemented in a usable fashion.

**Comment by aktentasche** [comment]
Yeah I guess this area is quite heterogenous. 

I would suggest that this feature then should set the 5 star photoprism internal rating on import, based on the metadata either embedded in the image file or some metadata file (e.g. my .xmp). And like you said, users should be able to set which "star source" should be the first to take if there are several. This would also solve the issue of a potential mismatch of star ratings in different sources on import (e.g. camera-set vs .xmp rating set in 3rd party software). A warning should still be logged though.

What do you think @graciousgrey? Happy to help with this new feature.

**Comment by graciousgrey** [comment]
We plan to implement this similarly to how we have already implemented it for other metadata fields.

- https://docs.photoprism.app/user-guide/library/metadata/
- https://www.photoprism.app/kb/metadata

During indexing, we read the available metadata from EXIF or XMP. The priority of different fields is defined in the source code. If the EXIF or XMP data of the original file is updated later, the change can be read on re-indexing, provided that the file's modification date has been updated and there has not been a manual change in PhotoPrism (which would have a higher priority). 

Writing metadata that has been added or edited in PhotoPrism back to EXIF or XMP is a different topic. We plan to do this, but we don't know when we will get to it yet. Currently, our development focus is on batch editing and AI enhancements.

**Comment by aktentasche** [comment]
Hi, I have been working on a script that populates an album based on ratings found in the XMP files so I have something until this feature is released. Maybe it is of help to someone: https://github.com/aktentasche/photoprism-tools/blob/main/populate_album_with_4_5_star_xmps_and_favorites.py

**Comment by ameyypawar** [comment]
I have a working implementation of this on my fork, built along the lines @graciousgrey described last August: ratings are read from Exif/XMP `Rating` during indexing and go through the usual source priority, so anything set in PhotoPrism itself always wins over a re-import. Photos explicitly rated 0 stars stay distinguishable from photos that were never rated. It also adds a `rating:` search filter (minimum stars, map view included) and star editing in the photo dialog. I left batch editing and metadata write-back out on purpose given the ongoing work in those areas.

Diff: https://github.com/photoprism/photoprism/compare/develop...ameyypawar:photoprism:feature/713-star-rating-v2

Would you be open to a PR for this? No hard feelings if you'd rather keep it in-house — just let me know either way.

**Comment by graciousgrey** [comment]
@ameyypawar Thanks for putting this together! We'd absolutely welcome a PR. Please go ahead and open one, and we'll review it as soon as our time allows.

From a first look it fits our approach nicely, reading Rating through the source priority so anything set in PhotoPrism wins over a re-import is exactly how we'd want it, and keeping explicit 0-star distinguishable from "never rated".

You're right that writing ratings back to Exif/XMP should stay out of scope for now as metadata write-back is something we want to handle consistently across all fields once our team has implemented it.

We can look at adding the rating to the batch edit dialog and to the photo viewer sidebar as part of reviewing/merging the PR, so no need to worry about those on your end unless you'd like to.
